packages/hash/external-services/docker-compose.yml

```
@@ -32,31 +36,76 @@ services:
     #   - ../../../var/external-services/opensearch/data:/usr/share/opensearch/data
 
   postgres:
-    image: postgres:14-alpine
+    image: postgres:14-alpine3.16
     deploy:
       restart_policy:
         condition: on-failure
-    env_file:
-      - ../../../.env
-      - ../../../.env.local
     environment:
-      - PGDATA=/data/pgdata
+      PGDATA: /data/pgdata
+      POSTGRES_USER: "${POSTGRES_USER}"
+      POSTGRES_HOST: "${POSTGRES_HOST}"
+      POSTGRES_PASSWORD: "${POSTGRES_PASSWORD}"
+      POSTGRES_PORT: "${POSTGRES_PORT}"
+      HASH_KRATOS_PG_USER: "${HASH_KRATOS_PG_USER}"
+      HASH_KRATOS_PG_PASSWORD: "${HASH_KRATOS_PG_PASSWORD}"
+      HASH_KRATOS_PG_DATABASE: "${HASH_KRATOS_PG_DATABASE}"
+      HASH_GRAPH_PG_USER: "${HASH_GRAPH_PG_USER}"
+      HASH_GRAPH_PG_PASSWORD: "${HASH_GRAPH_PG_PASSWORD}"
+      HASH_GRAPH_PG_DATABASE: "${HASH_GRAPH_PG_DATABASE}"
     ports:
       - "${POSTGRES_PORT}:5432"
     volumes:
-      - ../../../var/external-services/postgres/data:/data
+      - hash-postgres-data:/var/lib/postgresql/data
       - ./postgres/postgresql.conf:/etc/postgresql/postgresql.conf:ro
-      - ./postgres/init-user-db.sh:/docker-entrypoint-initdb.d/init-user-db.sh
+      - ./postgres/init-user-db.sh:/docker-entrypoint-initdb.d/init-user-db.sh:ro
+    healthcheck:
+      test: ["CMD-SHELL", "pg_isready --username ${POSTGRES_USER}"]
+      interval: 2s
+      timeout: 2s
+      retries: 5
+    command: -c 'config_file=/etc/postgresql/postgresql.conf'
+
+  graph:
+    init: true
+    depends_on:
+      postgres:
+        condition: service_healthy
+    build:
+      dockerfile: deployment/graph/Dockerfile
+      context: ../../graph
+      args:
+        PROFILE: dev
+    environment:
+      HASH_GRAPH_USER: "${HASH_GRAPH_PG_USER}"
+      HASH_GRAPH_PASSWORD: "${HASH_GRAPH_PG_PASSWORD}"
+      HASH_GRAPH_HOST: "postgres"
+      HASH_GRAPH_API_PORT: "5432"
```

I think we should use the existing environment variables for these:
```suggestion
      HASH_GRAPH_HOST: ${POSTGRES_HOST}
      HASH_GRAPH_API_PORT: ${POSTGRES_PORT}
```

---

That's not possible, they are picked based on the docker environment. `"postgres"` refers to the other service and `"5432"` is a hard-coded port. `"${POSTGRES_PORT}"` is the port exposed from docker.

---

Good point on the host having to be `"postgres"`, but isn't `"${POSTGRES_PORT}"` the port exposed from the docker container? i.e. any other container would have to use the value of `"${POSTGRES_PORT}"` to connect to it?

My understanding is that this isn't hardcoded, and is set [on line 56 of this file](https://github.com/hashintel/hash/blob/3c64a7c04cd1c55a1ee9304928f2dbb58f2a0a2c/packages/hash/external-services/docker-compose.yml#L56)

---

No, it's not set in 56. `POSTGRES_PORT` and `POSTGRES_HOST` is visible for the host network, it's forwarding the port from the docker network. However, the Graph is not in the host network but in the compose network and one has to specify the network for the postgres service (`"postgres"`). When specify a different `POSTGRES_PORT` and also setting this here, the migration script is failing, this needs to be postgres:5432. Anyway, I another weird error when using a different port: I cannot connect to the REST API for the Graph at all. Getting the OpenAPI specs is working.